### PR TITLE
Semantic model tuple types

### PR DESCRIPTION
**Pull request by chsienki**
Correctly report nullability of tuples in the semantic model: 
- Store the original tuple literal rather than just its type
- Visit the original literal without diagnostics only in the nullability walker to understand its type
- Update the generator to allow force visiting fields in the nullable re-writer
- Update various tuple tests to be correct

Related: #35010

### CODE REVIEW

**Comment by chsienki** [comment]
@dotnet/roslyn-compiler for review please

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 1]
```diff
@@ -1527,8 +1527,8 @@
   <Node Name="BoundConvertedTupleLiteral" Base="BoundTupleExpression">
     <!-- Converted tuple must have a type -->
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
-    <!-- Natural type is preserved for the purpose of semantic model. Can be null -->
-    <Field Name="NaturalTypeOpt" Type="TypeSymbol" Null="allow"/>
+    <!-- Original tuple is preserved for the purpose of semantic model -->
+    <Field Name="SourceTuple" Type="BoundTupleLiteral" SkipInVisitor="true" ForceInNullabilityRewriter="true" />
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: >ForceInNullabilityRewriter [](start = 76, length = 26)

Perhaps `UpdateFieldInNullabilityRewriter` instead, so it doesn't sound like it conflicts with `SkipInNullabilityRewriter`

Comment (333fred) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: Or perhaps we can update `SkipInVisitor` to accept more than just true/false

---
In reply to: [296056473](https://github.com/dotnet/roslyn/pull/36628#discussion_r296056473) [](ancestors = 296056473)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -3986,6 +3986,13 @@ public override BoundNode VisitTupleLiteral(BoundTupleLiteral node)
 
         public override BoundNode VisitConvertedTupleLiteral(BoundConvertedTupleLiteral node)
         {
+            // Visit the source tuple so that the semantic model can correctly report nullability for it
+            // Disable diagnostics, as we don't want to duplicate any that are produced by visiting the actual literal below
+            var previousDisableDiagnostics = _disableDiagnostics;
+            _disableDiagnostics = true;
+            VisitTupleExpression(node.SourceTuple);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Can we use `VisitWithoutDiagnostics` instead?

**Comment by 333fred** [comment]
Can you please add some tests of the public API to show that this is correctly showing the differences between `Nullability` and `ConvertedNullability`?

**Comment by 333fred** [comment]
Done review pass (commit 2)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -52154,14 +52154,12 @@ void M()
...
         {
-            // Type of (default, default) should not be <null>!. Was (object?, string?)
-            // https://github.com/dotnet/roslyn/issues/35010
             var source =
 @"class Program
 {
     static void F()
     {
-        (object? a, string) t = (default, default)/*T:<null>!*/; // 1
+        (object? a, string) t = (default(object), default(string)) /*T:(object?, string?)*/; // 1
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >(default(object), default(string)) [](start = 32, length = 34)

Is it necessary to change the test?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -52198,30 +52196,28 @@ static void F()
...
     {
-        (object x, object? y, object? z) t = (null, x, y)/*T:<null>!*/; // 1
+        (object x, object? y, object? z) t = (null, x, y); // 1
         _ = t/*T:(object! x, object? y, object? z)*/;
         t.x.ToString(); // 2
         t.y.ToString(); // 3
         t.z.ToString();
         if (x == null) return;
-        (object x, object y) u = (x, default)/*T:<null>!*/; // 4
+        (object x, object y) u = (x, default(object))/*T:(object! x, object?)*/; // 4
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >default(object) [](start = 37, length = 15)

Is it necessary to change the test?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 5]
```diff
@@ -52543,8 +52528,8 @@ static void F(bool b, object x)
             var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
             comp.VerifyDiagnostics(
                 // (6,13): warning CS8619: Nullability of reference types in value of type '(object, object, object, object?, object?, (object, object?), object, object, object?, object?)' doesn't match target type '(object?, object, object?, object, object?, (object, object?), object, object, object?, object)'.
-                //             (x, x, x, default, default, default, x, x, default, default)/*T:<null>!*/; // 1
-                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "(x, x, x, default, default, default, x, x, default, default)").WithArguments("(object, object, object, object?, object?, (object, object?), object, object, object?, object?)", "(object?, object, object?, object, object?, (object, object?), object, object, object?, object)").WithLocation(6, 13),
+                //             (x, x, x, default(object), default(object), default((object, object?)), x, x, default(object), default(object))/*T:(object!, object!, object!, object?, object?, (object!, object?), object!, object!, object?, object?)*/; // 1
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >default(object) [](start = 41, length = 15)

Same question.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 6]
```diff
@@ -53105,6 +53090,66 @@ static void G(object? x, object? y)
...
+        {
+            var source =
+@"
+#pragma warning disable CS0219
+class C
+{
+    static void F(string? x)
+    {
+     //   C/*T:C!*/ a = x/*T:string?*/;
+        //(C, C) a = (new C(), new C());
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Should these lines be uncommented or removed?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 3 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Yes. `(default, default)` doesn't actually have a type, it's target typed from the LHS. Changing the tests allows us to check that we're correctly returning the type we expect to.

---
In reply to: [296373455](https://github.com/dotnet/roslyn/pull/36628#discussion_r296373455) [](ancestors = 296373455)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 6 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Mistakenly included, will remove, thanks :)

---
In reply to: [296374978](https://github.com/dotnet/roslyn/pull/36628#discussion_r296374978) [](ancestors = 296374978)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 3 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Will add to the test where we use target typing too

---
In reply to: [296394150](https://github.com/dotnet/roslyn/pull/36628#discussion_r296394150) [](ancestors = 296394150,296373455)

**Comment by chsienki** [comment]
Added NullablePublicAPITests.TupleAssignment and updated the type checker to allow checking for converted type too.

---
In reply to: [504242517](https://github.com/dotnet/roslyn/pull/36628#issuecomment-504242517) [](ancestors = 504242517)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 7]
```diff
@@ -105312,7 +105438,7 @@ class Program
             comp.VerifyDiagnostics();
         }
 
-        // C<T~> operator op(C<T!>, C<T~>) where T : class
+        // C<T~> operator op(C<T!>, C<T~>) where T : class 2
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: > 2 [](start = 58, length = 2)

Is this change intentional?

Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Hah. No. Oh man am I ready for the weekend apparently. Thanks

**Comment by gafter** [comment]
I am changing the shape of `BoundConvertedTupleLiteral` in my pending work.  I am not sure how much of a conflict this will cause.

---
In reply to: [504223575](https://github.com/dotnet/roslyn/pull/36628#issuecomment-504223575) [](ancestors = 504223575)

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 8]
```diff
@@ -1527,8 +1527,8 @@
   <Node Name="BoundConvertedTupleLiteral" Base="BoundTupleExpression">
     <!-- Converted tuple must have a type -->
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
-    <!-- Natural type is preserved for the purpose of semantic model. Can be null -->
-    <Field Name="NaturalTypeOpt" Type="TypeSymbol" Null="allow"/>
+    <!-- Original tuple is preserved for the purpose of semantic model -->
+    <Field Name="SourceTuple" Type="BoundTupleLiteral" SkipInVisitor="ExceptNullabilityRewriter" />
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: >SourceTuple [](start = 17, length = 11)

I will have to make this field optional when merging, as nodes of `BoundConvertedTupleLiteral` need to be created when lowering tuple conversions and tuple equality.  We currently generate `BoundTupleLiteral`, but that isn't correct as such nodes should be gone by the end of initial binding and we don't want to have to deal with both forms.
